File: messagematch/src/main/java/org/forwoods/messagematch/match/PathMatcher.java
package org.forwoods.messagematch.match;

import org.forwoods.messagematch.match.fieldmatchers.FieldMatcher;
import org.forwoods.messagematch.spec.URIChannel;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class PathMatcher
{
    private final String matcherPath;
    private final String concretPath;
    private final Map<String, Object> bindings;

    public PathMatcher(final String matcherPath, final String concretePath, final Map<String, Object> inputBindings)
    {
        this.matcherPath = matcherPath;
        this.bindings = new HashMap<>(inputBindings);
        this.concretPath = concretePath;
    }

    public boolean matches()
    {
        PathMatcherPattern pattern = new PathMatcherPattern(matcherPath);
        final Matcher matcher = pattern.pattern.matcher(concretPath);
        if (matcher.matches()) {
            List<String> bindingExpressions = pattern.bindingExpressions;
            for (int i = 0; i < bindingExpressions.size(); i++)
            {
                final String bindingExpression = bindingExpressions.get(i);
                FieldMatcher<?> fieldMatcher = JsonMatcher.parseMatcher(bindingExpression);
                String concreteValue = matcher.group(i+1);
                final boolean matches = fieldMatcher.matches(concreteValue, bindings);
                if (!matches) {
                    return false;
                }
            }
            return true;
        }
        return false;
    }

    public Map<String, Object> getBindings()
    {
        return bindings;
    }

    private static class PathMatcherPattern {
        private final Pattern pattern;
        private final List<String> bindingExpressions = new ArrayList<>();
        private final static Pattern matcherDeconstruct = Pattern.compile("\\{[^}]+\\}");
        public PathMatcherPattern(String matcherPath) {
            StringBuilder patternBuilder = new StringBuilder();
            Matcher matcher = matcherDeconstruct.matcher(matcherPath);
            int pos = 0;
            while (matcher.find()) {
                patternBuilder.append(matcherPath.substring(pos, matcher.start()));
                patternBuilder.append("([^/]+)");
                final String group = matcher.group();
                bindingExpressions.add(group.substring(1, group.length()-1));
                pos = matcher.end();
            }
            patternBuilder.append(matcherPath.substring(pos));
            pattern = Pattern.compile(patternBuilder.toString());
        }
    }
}

File: messagematch-typescript/src/match/PathMatcher.ts
import {JsonMatcher} from "./JsonMatcher";

export class PathMatcher {
    constructor(private matcherPath:string, private concretePath:string, readonly binding:Record<string, any>) {
    }

    public matches():boolean {
        const pattern = new PathMatcherPattern(this.matcherPath);
        const matcher = this.concretePath.match(pattern.pattern)
        if (matcher) {
            for (let i = 0; i < pattern.bindingExpressions.length; i++) {
                const bindingExpression = pattern.bindingExpressions[i];
                const fieldMatcher = JsonMatcher.parseMatcher(bindingExpression);
                const concrete = matcher[i+1];
                if (!fieldMatcher.matches(concrete, this.binding)) return false;
            }
            return true;
        }
        return false;
    }
}

const matcherDeconstruct = /\{[^}]+\}/g;

class PathMatcherPattern {
    bindingExpressions:string[] =[];
    pattern:RegExp
    constructor (private matcherPath:string){
        let r:string='^';
        const pathbindings = matcherPath.matchAll(matcherDeconstruct)
        let pos = 0;
        for (const pathBinding of pathbindings){
            r+=matcherPath.substring(pos, pathBinding.index);
            r+="([^/]+)";
            this.bindingExpressions.push(pathBinding[0].substring(1, pathBinding[0].length-1));
            pos = pathBinding.index + pathBinding[0].length
        }
        r+=matcherPath.substring(pos);
        r+='$'
        this.pattern = new RegExp(r);
    }
}